package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,342 @@
+        DistMatrix : DistMatrix object
+            Distance matrix to be made into a diffusion kernel and perform
+            an eigenvector decomposition on.
+        epsilon : function, optional
+            Specifies the method used for the choice of scale parameter in the
+            diffusion map. More information in [1] and [2]. With 'average'
+            the average of the RMSD to the k-th greatest value is used.
+        weights: list, optional
+            The list has to have the same length as the trajectory.
+            With 'None' the weight of each frame of the trajectory will be the
+            same.
+        """
+        self._epsilon = epsilon
+        # this should be a reference to the same object as
+        # epsilon.scaledMatrix
+        self.dist_matrix = DistMatrix.dist_matrix
+        if weights_ker is None:
+            # weights do not apply to metric
+            self._weights_ker = np.ones((self.nframes,))
+        else:
+            if weights.shape[0] != self.nframes:
+                raise ValueError("The weight should have the same length as "
+                                 'the trajectory')
+            else:
+                # weights are constructed as relative to the mean
+                self._weights_ker = (np.asarray(weights, dtype=np.float64) /
+                                     np.mean(weights))
+
+    def _conclude(self):
+        self._epsilon.determineEpsilon()
+        self._kernel = self._epsilon.scaledMatrix
+        # take negative exponent of scaled matrix to create Isotropic kernel
+        self._kernel = np.exp(-self._kernel)
+
+        # define an anistropic diffusion term q
+        q_vector = np.zeros((self.nframes, ))
+        # weights should reflect the density of the points on the manifold
+        for i in range(self.nframes):
+            q_vector[i] = np.dot(self._kernel2[i, :], self._weights_ker)
+
+        # Form a new kernel from the anisotropic diffusion term q
+        self._kernel /= np.sqrt(q_vector[:, np.newaxis].dot(q_vector[np.newaxis]))
+
+        # Weighted Graph Laplacian normalization on this new graph
+        d_vector = np.zeros((self.nframes, ))
+        for i in range(self.nframes):
+            d_vector[i] = np.dot(self._kernel2[i, :], self._weights_ker)
+
+        for i in range(self.nframes):
+            self._kernel[i, :] = self._kernel2[i, :] * self._weights_ker
+
+        # Define anisotropic transition by dividing kernel by this term
+        self._kernel /= np.sqrt(d_vector[:, np.newaxis].dot(d_vector[np.newaxis]))
+
+        # Apply timescaling, HELP, I think I'm pythoning wrong
+        for i in range(t):
+            if i > 1:
+                self._kernel.__matmul__(self._kernel)
+
+        eigenvals, eigenvectors = np.linalg.eig(self._kernel)
+
+        eg_arg = np.argsort(eigenvals)
+        self.eigenvalues = eigenvals[eg_arg[::-1]]
+        self.eigenvectors = eigenvectors[eg_arg[::-1], :]
+
+    def spectral_gap(self):
```

Comment:
Find where the eigenvalues fall off significantly. The kernel should have
some number of dominant eigenvalues reflecting the intrinsic dimensionality
of the system.
On Thu, Jun 16, 2016 at 12:53 AM kain88-de <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r67301245>:
>
> > +
> > +        # Define anisotropic transition by dividing kernel by this term
> > +        self._kernel /= np.sqrt(d_vector[:, np.newaxis].dot(d_vector[np.newaxis]))
> > +
> > +        # Apply timescaling, HELP, I think I'm pythoning wrong
> > +        for i in range(t):
> > +            if i > 1:
> > +                self._kernel.__matmul__(self._kernel)
> > +
> > +        eigenvals, eigenvectors = np.linalg.eig(self._kernel)
> > +
> > +        eg_arg = np.argsort(eigenvals)
> > +        self.eigenvalues = eigenvals[eg_arg[::-1]]
> > +        self.eigenvectors = eigenvectors[eg_arg[::-1], :]
> > +
> > +    def spectral_gap(self):
>
> What will this function do?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/fc70b11cb43720949f7fad2bc7ae05211c5acea8#r67301245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARpWEzo5CyTf36lgqOANoqwo2lNVbks5qMQEMgaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs